ITensorGaussianMPS/src/gmps.jl

```diff
@@ -391,6 +817,103 @@ function interleave(xs...)
   return res
 end
 
+function interleave(M::AbstractMatrix)
+  @assert size(M, 1) == size(M, 2)
+  n = div(size(M, 1), 2)
+  first_half = Vector(1:n)
+  second_half = Vector((n + 1):(2 * n))
+  interleaved_inds = interleave(first_half, second_half)
+  return M[interleaved_inds, interleaved_inds]
+end
+
+function interleave(g1::Circuit, g2::Circuit)
+  return Circuit(interleave(g1.rotations, g2.rotations))
+end
+
+function reverse_interleave(M::AbstractMatrix)
+  @assert size(M, 1) == size(M, 2)
+  n = div(size(M, 1), 2)
+  first_half = Vector(1:n)
+  second_half = Vector((n + 1):(2 * n))
+  interleaved_inds = interleave(first_half, second_half)
+  ordered_inds = sortperm(interleaved_inds)
+  return M[ordered_inds, ordered_inds]
+end
+
+function correlation_matrix_to_mps(
+  s::Vector{<:Index},
+  Λ_up0::AbstractSymmetry,
+  Λ_dn0::AbstractSymmetry;
+  eigval_cutoff::Float64=1e-8,
+  maxblocksize::Int=min(size(Λ_up0, 1), size(Λ_dn0, 1)),
+  kwargs...,
+)
+  @assert size(Λ_up.data, 1) == size(Λ_up.data, 2)
+  @assert size(Λ_dn.data, 1) == size(Λ_dn.data, 2)
+  Λ_up = maybe_drop_pairing_correlations(Λ_up0)
+  Λ_dn = maybe_drop_pairing_correlations(Λ_dn0)
+  #@assert Λ_up<:aT && Λ_dn<:aT where aT<:AbstractSymmetry  ##ToDo:implement this check
+  #calctype = typeof(Λ_up)
+  N_up = size(Λ_up.data, 1)
+  N_dn = size(Λ_dn.data, 1)
+  N = N_up + N_dn
+  ns_up, C_up = correlation_matrix_to_gmps(
+    Λ_up; eigval_cutoff=eigval_cutoff, maxblocksize=maxblocksize
+  )
+  ns_dn, C_dn = correlation_matrix_to_gmps(
+    Λ_dn; eigval_cutoff=eigval_cutoff, maxblocksize=maxblocksize
+  )
+  C_up = mapindex(n -> 2n - 1, C_up)
+  C_dn = mapindex(n -> 2n, C_dn)
+  if Λ_up <: ConservesNfParity && Λ_dn <: ConservesNfParity
+    C = Circuit(
+      interleave(
+        interleave(C_up.rotations[1:2:end], C_dn.rotations[1:2:end]),
+        interleave(C_up.rotations[2:2:end], C_dn.rotations[2:2:end]),
+      ),
+    )
+    ns = interleave(
+      interleave(ns_up[1:2:end], ns_dn[1:2:end]), interleave(ns_up[2:2:end], ns_dn[2:2:end])
+    )
+  elseif Λ_up <: ConservesNf && Λ_dn <: ConservesNf
+    C = Circuit(interleave(C_up.rotations, C_dn.rotations))
+    ns = interleave(ns_up, ns_dn)
+  else
+    error("Λ_up and Λ_dn have incompatible subtypes of AbstractSymmetry")
+  end
```

Comment:
This looks like a good candidate for splitting this off by overloading `interleave(::AbstractSymmetry, ::AbstractSymmetry)`.

Reply:
Agreed. I'll address it when the spinful case is implemented properly.